### PR TITLE
Add comprehensive test suite for routing and frontend services

### PR DESCRIPTION
**Pull request by brunobuddy**
## Summary
This PR adds extensive unit test coverage for the routing system, OAuth services, proxy adapters, and frontend components. The test suite covers core routing logic, provider management, model discovery, caching, and API integrations.

### Backend Routing Tests
- **ResolveService** (`resolve.service.spec.ts`): Tests for model resolution via specificity assignments, tier scoring, provider authentication, and fallback logic
- **TierAutoAssignService** (`tier-auto-assign.service.spec.ts`): Tests for automatic tier assignment based on model capabilities and pricing
- **RoutingCacheService** (`routing-cache.service.spec.ts`): Tests for caching tiers, providers, custom providers, and specificity assignments with TTL expiration
- **ResolveAgentService** (`resolve-agent.service.spec.ts`): Tests for agent resolution and tenant validation
- **CustomProviderService** (`custom-provider.service.spec.ts`): Tests for CRUD operations on custom providers with URL validation and cache invalidation

### OAuth & Authentication Tests
- **OpenaiOauthService** (`openai-oauth.service.spec.ts`): Tests for OAuth flow initialization, token exchange, and model discovery
- **MinimaxOauthService** (`minimax-oauth.service.spec.ts`): Tests for Minimax OAuth with region-specific endpoints and device code polling
- **MinimaxOauthHelpers** (`minimax-oauth-helpers.spec.ts`): Tests for URL builders and region validation
- **CopilotDeviceAuthService** (`copilot-device-auth.service.spec.ts`): Tests for GitHub device code flow

### Proxy & Adapter Tests
- **ChatGPT Adapter** (`chatgpt-adapter.spec.ts`): Tests for converting between OpenAI and Responses API formats, including tool calls and streaming
- **ChatGPT Helpers** (`chatgpt-helpers.spec.ts`): Tests for message conversion, tool handling, and SSE formatting
- **Cache Injection** (`cache-injection.spec.ts`): Tests for OpenRouter cache control header injection

### Utility Tests
- **SQL Dialect** (`sql-dialect.spec.ts`): Tests for database dialect detection and SQL generation helpers

### Frontend Tests
- **ProviderApiKeyTab** (`ProviderApiKeyTab.test.tsx`): Tests for provider list sorting, local-only provider handling, and connection status
- **ChartCard** (`ChartCard.test.tsx`): Tests for chart view switching and stat formatting
- **CostByModelTable** (`CostByModelTable.test.tsx`): Tests for table rendering and cost-based sorting
- **EmailProviderSection** (`EmailProviderSection.test.tsx`): Tests for setup/edit/remove flows
- **AuthBadge** (`AuthBadge.test.tsx`): Tests for authentication type labeling

### API Tests
- **API Extra** (`api-extra.test.ts`): Tests for free models, OAuth callbacks, and specificity management endpoints

### Shared Tests
- **Agent Type** (`agent-type.spec.ts`): Tests for agent category/platform definitions and mappings

## Notable Implementation Details
- Mock factories used throughout to reduce boilerplate and improve test maintainability
- Comprehensive edge case coverage including null/undefined handling, fallback logic, and error scenarios
- Fake timers used for TTL and expiration testing
- Mocked external dependencies (fetch, config services) to isolate units under test
- Tests verify both happy paths and error conditions for OAuth flows and provider operations

https://claude.ai/code/session_01UtDok1xHxoRkmsetvzFbRU

---
## Summary by cubic
Adds a comprehensive test suite for backend routing, OAuth providers, proxy adapters, SQL/utility helpers, and frontend components/services. Also adds isolated tests for the `current-user` decorator and fixes test cleanup to prevent global leaks.

- **Bug Fixes**
  - Import `reflect-metadata` in `current-user.decorator.spec.ts` so it runs in isolation.
  - Use `vi.unstubAllGlobals()` in `api-extra.test.ts` cleanup to undo `vi.stubGlobal` and avoid leaking `fetch`/`location`.

<sup>Written for commit 27b347bb113e0419866c263eb55a1816bc1482ef. Summary will update on new commits.</sup>

### CODE REVIEW

File: packages/backend/src/routing/routing-core/tier-auto-assign.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+      const existing: Partial<TierAssignment>[] = [
+        { tier: 'simple', agent_id: 'agent-1', auto_assigned_model: null },
+      ];
+      const { svc, save, insert } = makeService({ models, existingTiers: existing });
+      await svc.recalculate('agent-1');
+
+      expect(save).toHaveBeenCalledTimes(1);
+      expect(save.mock.calls[0][0]).toHaveLength(1);
+      expect(save.mock.calls[0][0][0].tier).toBe('simple');
+      expect(save.mock.calls[0][0][0].auto_assigned_model).toBe('openai/gpt-5');
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/routing-core/tier-auto-assign.service.spec.ts:212]: Neither model in this test is a subscription model — both default to `authType: 'api_key'` in the local `m()` helper (the helper sets `authType: 'api_key'` before the `...partial` spread, and neither `m({ ... })` call here overrides it). So `subModels` is `[]`, and the simple tier falls through to `pickBest(keyModels, 'simple')`, which picks the cheapest by total price: `openai/gpt-5` (5+5=10) beats `anthropic/claude-opus-4` (10+10=20). Changing the assertion to `anthropic/claude-opus-4` would make the test fail.

Subscription-first routing is exercised separately by the "prefers subscription models over API-key models for every tier" test below, which explicitly sets `authType: 'subscription'`. Leaving this as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_
